App/Sagas/ImageSharingTriggers.ts

```diff
@@ -0,0 +1,78 @@
+import { insertImage, addToIpfs, uploadArchive, shareWalletImage, addToWallet, shareToThread } from './ImageSharingSagas'
+
+export function * handleSharePhotoRequest (action: ActionType<typeof UIActions.sharePhotoRequest>) {
+  const { image, threadId, comment } = action.payload
+  if (!image || !threadId) {
+    return
+  }
+  if (typeof image === 'string') {
+    yield call(shareWalletImage, image, threadId, comment)
+  } else {
+    yield call(insertImage, image, threadId, comment)
+  }
+}
+
+export function * handleImageUploadComplete (action: ActionType<typeof ProcessingImagesActions.imageUploadComplete>) {
+  const { uuid } = action.payload
+  yield call(addToWallet, uuid)
+  try {
+    const processingImage: ProcessingImage | undefined = yield select(ProcessingImagesSelectors.processingImageByUuid, uuid)
+    if (processingImage && processingImage.addData && processingImage.addData.addResult.archive) {
+      const exists: boolean = yield call(RNFS.exists, processingImage.addData.addResult.archive.path)
+      if (exists) {
+        yield call(RNFS.unlink, processingImage.addData.addResult.archive.path)
+      }
+    }
+  } catch (e) {}
+}
+
+export function * retryImageShare (action: ActionType<typeof ProcessingImagesActions.retry>) {
+  const { uuid } = action.payload
+  const processingImage: ProcessingImage | undefined = yield select(ProcessingImagesSelectors.processingImageByUuid, uuid)
+  if (!processingImage) {
+    return
+  }
+  if (processingImage.state === 'addedToWallet' || processingImage.state === 'sharing') {
+    yield call(shareToThread, uuid)
+  } else if (processingImage.state === 'uploaded' || processingImage.state === 'addingToWallet') {
+    yield call(addToWallet, uuid)
+  } else if (processingImage.state === 'added' || processingImage.state === 'uploading') {
+    yield call(uploadArchive, uuid)
+  } else if (processingImage.state === 'pending' || processingImage.state ==='adding') {
+    yield call(addToIpfs, uuid)
+  }
+}
+
+export function * cancelImageShare (action: ActionType<typeof ProcessingImagesActions.cancelRequest>) {
+  const { uuid } = action.payload
+  try {
+    const processingImage: ProcessingImage | undefined = yield select(ProcessingImagesSelectors.processingImageByUuid, uuid)
+    if (!processingImage) {
+      return
+    }
+
+    // Delete the shared image if needed
+    const exists: boolean = yield call(RNFS.exists, processingImage.sharedImage.path)
+    if (exists && processingImage.sharedImage.canDelete) {
+      yield call(RNFS.unlink, processingImage.sharedImage.path)
+    }
+
+    // Delete the Textile payload
+    if (processingImage.addData && processingImage.addData.addResult.archive) {
+      const exists: boolean = yield call(RNFS.exists, processingImage.addData.addResult.archive.path)
+      if (exists) {
+        yield call(RNFS.unlink, processingImage.addData.addResult.archive.path)
+      }
+    }
+
+    // What else? Undo local add, remote pin, remove from wallet?
```

Comment:
Ok, I added code to delete the image and upload archive. What else can we do here?

Reply:
I don't display the cancel button unless the share has failed. Another option is to always display the cancel button to allow canceling of in-flight shares. That might complicate this more, not sure. Thoughts appreciated.

Reply:
@andrewxhill and @sanderpick this is still WIP mostly because of this question.

Reply:
+1 for the way you did it

Reply:
- If it failed _after_ being added to IPFS, we should eventually unpin from local node (not all of it is pinned to begin with, but still would be good to unpin the thumb, small, and meta). I guess this would only happen if the add / share to thread failed. I think skippable for now... let's see what breaks first :)

Reply:
ok cool.